### PR TITLE
Polish HostMessages

### PR DESCRIPTION
**Pull request by abarth**
Cleans up a few style nits in hello_services and adds support for automatic
JSON encoding and decoding to the HostMessages interface.

### CODE REVIEW

**Comment by abarth** [comment]
@jason-simmons @Hixie

File: examples/hello_services/lib/main.dart [hunk 1]
```diff
@@ -50,14 +47,14 @@ class _HelloServicesState extends State<HelloServices> {
...
-  void _getLocation() {
+  Future<Null> _getLocation() async {
     Map<String, String> message = <String, String>{'provider': 'network'};
-    HostMessages.sendToHost('getLocation', JSON.encode(message))
-        .then(_onReceivedLocation);
-  }
-
-  void _onReceivedLocation(String json) {
-    Map<String, num> reply = JSON.decode(json);
+    Map<String, num> reply = await HostMessages.sendJSON('getLocation', message);
```
Comment (Hixie) [inline examples/hello_services/lib/main.dart]: won't the returned Map be a `Map<String, dynamic>` ?

File: packages/flutter/lib/src/services/host_messages.dart [hunk 2]
```diff
@@ -48,13 +49,28 @@ final _ApplicationMessagesImpl _appMessages = new _ApplicationMessagesImpl();
...
-  static Future<String> sendToHost(String messageName, String message) async {
+  static Future<String> sendToHost(String messageName, [String message = '']) async {
     return (await _hostAppMessagesProxy.sendString(messageName, message)).reply;
   }
 
-  /// Register a callback for messages received from the host application.
-  /// The callback function must return a String, Future<String>, or null.
+  /// Sends a JSON-encoded message to the host application and JSON-decodes the response.
+  static Future<dynamic> sendJSON(String messageName, [dynamic json]) async {
+    return JSON.decode((await _hostAppMessagesProxy.sendString(messageName, JSON.encode(json) ?? '')).reply);
```
Comment (Hixie) [inline packages/flutter/lib/src/services/host_messages.dart]: `JSON.encode(json)` can't return null as far as I can tell

File: examples/hello_services/lib/main.dart [hunk 3]
```diff
@@ -3,23 +3,20 @@
...
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
 
 final Random random = new Random();
 
-Future<String> handleGetRandom(String json) async {
-  Map<String, dynamic> message = JSON.decode(json);
-  double min = message['min'].toDouble();
-  double max = message['max'].toDouble();
+Future<dynamic> handleGetRandom(Map<String, dynamic> message) async {
```
Comment (Hixie) [inline examples/hello_services/lib/main.dart]: this argument is of type `dynamic`

File: examples/hello_services/lib/main.dart [hunk 4]
```diff
@@ -3,23 +3,20 @@
...
+  final double min = message['min'].toDouble();
+  final double max = message['max'].toDouble();
 
-  double value = (random.nextDouble() * (max - min)) + min;
-
-  Map<String, double> reply = <String, double>{'value': value};
-  return JSON.encode(reply);
+  return <String, double>{
+    'value': (random.nextDouble() * (max - min)) + min
+  };
```
Comment (Hixie) [inline examples/hello_services/lib/main.dart]: why not just return a raw double?

File: examples/hello_services/lib/main.dart [hunk 5]
```diff
@@ -50,14 +47,14 @@ class _HelloServicesState extends State<HelloServices> {
     );
   }
 
-  void _getLocation() {
+  Future<Null> _getLocation() async {
     Map<String, String> message = <String, String>{'provider': 'network'};
```
Comment (Hixie) [inline examples/hello_services/lib/main.dart]: final

**Comment by Hixie** [comment]
LGTM.
ObjectiveC changes RSLGTM.

**Pull request merged**